Add Tailscale as Third Party Integration for Palette Edge (#1878)
spectrocloud/librarium: docs/docs-content/clusters/edge/third-party-integrations/tailscale.md

        commands:
          - |
            ID=$(cat /sys/class/dmi/id/product_uuid)
            if [ -f /oem/tailscale/tailscaled.state ]; then
              systemctl stop tailscaled
              cp /oem/tailscale/tailscaled.state /var/lib/tailscale/tailscaled.state
              systemctl start tailscaled
              tailscale up --ssh --hostname="edge-${ID}"
            else
              tailscale up --authkey=$AUTH-KEY --ssh --hostname="edge-${ID}"
              mkdir /oem/tailscale
              cp /var/lib/tailscale/tailscaled.state /oem/tailscale/tailscaled.state
            fi
      - name: "Enable Tailscale"
        if: '[ ! -f "/run/cos/recovery_mode" ] && grep _current-profile /var/lib/tailscale/tailscaled.state'
        commands:
          - |
            ID=$(cat /sys/class/dmi/id/product_uuid)
            tailscale up --ssh --hostname="edge-${ID}"
  ```

  If you already have a `stages` block in your user-data file, you must merge the existing block together with the above content. The `stages` block is based on Kairos cloud-init stages. For more information on cloud init stages, refer to [Cloud Init Stages](../edge-configuration/cloud-init.md).

  :::info

>   In the above `stages` block, you are using the device ID of your Edge device (read from `/sys/class/dmi/id/product_uuid`) as the hostname with which to register your device with Tailscale. For more information about how this ID is generated, refer to [Install Configurations](../edge-configuration/installer-reference.md#device-id-uid-parameters).

Reviewer: ```suggestion
  In the above `stages` block, you are using the device ID of your Edge device that is read from the file **/sys/class/dmi/id/product_uuid**, as the hostname with which to register your device with Tailscale. For more information about how this ID is generated, refer to [Install Configurations](../edge-configuration/installer-reference.md#device-id-uid-parameters).
```

  If you want to use a different hostname, especially when using the `deviceUIDPaths` parameter in the user-data, you can adjust the two `ID=$(cat /sys/class/dmi/id/product_uuid)` lines in the content above to match your custom device naming configuration.

  :::

7. Build the Edge device installation ISO and providers images. Afterward, push the provider images to an image registry. For more information, refer to [Build Edge Artifacts](../edgeforge-workflow/palette-canvos.md).

> 8. Flash your external volume with the Edge installer ISO image. We recommend using [balena etcher](https://etcher.balena.io/) to flash your volume. 

Reviewer: Is there a specific reason for why we would recommend this specific tool versus just mentioning it as an example? In other Edge docs, we mention it as an option, but we stay away from tooling recommendations unless there is a good reason for it.

> 8. Flash your external volume with the Edge installer ISO image. We recommend using [balena etcher](https://etcher.balena.io/) to flash your volume. 

Reviewer 2: ```suggestion
8. Flash your external volume with the Edge installer ISO image. You can use [balena etcher](https://etcher.balena.io/) or any other tool of your choice to flash your volume. 
```

9. Plug the external volume into your Edge device and boot up the device using the volume to prepare your Edge device for installation. For more information, refer to [Prepare Edge Host for Installation](../site-deployment/stage.md). 

10. Remove the volume and boot up your device again to register your Edge host. If the Edge host has internet access, it will start up Tailscale and register your device with Tailscale. 

## Validate

1. Log in to [Tailscale console](https://login.tailscale.com/admin/machines). 

> 2. In the **Machines** tab, your Edge device is displayed in the Machines list. You can SSH to your host from any device that is also connected to your Tailscale network.

Reviewer: ```suggestion
2. In the **Machines** tab, your Edge device is displayed in the Machines list. You can SSH to your host from any device that is also connected to your Tailscale network. Check out the [Tailscale SSH](https://tailscale.com/kb/1193/tailscale-ssh) documentation page to learn more about SSH with Tailscale.
```
This could help newer users.

## Troubleshooting

### All Traffic Dropped for 100.64.0.0/10 CIDR Range 

> Tailscale uses the 100.64.0.0/10 range of IP addresses for your Tailnets. That means that by default, this address range (or parts of it) cannot be used for any of the following:

Reviewer: ```suggestion
Tailscale uses the 100.64.0.0/10 range of IP addresses for your Tailnets. That means that by default, this address range, or parts of it, cannot be used for any of the following:
```
We try to avoid parentheses inside sentences that convey information. The sentences can be rewritten to a more formal tone by omitting the paratheses. Paratheses are also a bit distracting.

- Kubernetes cluster pod CIDR
- Kubernetes cluster service CIDR
- Palette Edge Overlay network CIDR

#### Debug Steps

If you want to use parts of the 100.64.0.0/10 range for your Kubernetes clusters or your Palette Edge Overlay networks, you must limit the IP address range that your Tailnet uses to a fraction of the 100.64.0.0/10 range. Use the following steps to limit your Tailnet range:

1. First, configure an IP Pool in Tailscale. We have found the following configuration works well to assign addresses in the new range to all nodes:

> ```json

Reviewer: Let's clean up the white spacing on this code snippet.

Reviewer 2: hmm, not sure why it's showing up like that - doesn't have the white space in my code editor

	"nodeAttrs": [
		{
			"target": ["*"],
			"ipPool": ["100.74.0.0/16"],
		},
	],
```

2. Next, in the OS pack of your cluster profile, add the following:

> ```yaml {14}

Reviewer 2: @kreeuwijk Hi, I think this is basically ready to merge. Just one more question - any reason to highlight line 14? It sounds like the user need to add the entire block to their profile

Author: @lennessyy oh sorry I didn't know what `{14}` did, I thought it was some layout hint or something so I just copied it.

            Description=Tailscale iptables fix schedule
            [Timer]
            OnBootSec=15
            OnUnitActiveSec=15
            [Install]
            WantedBy=timers.target
    - name: "Tailscale adjustment script"
      files:
        - path: /etc/palette/tailscale-iptables.sh
          permissions: 0755
          owner: 0
          group: 0
          content: |
            #!/bin/sh
            if iptables -L ts-input | grep DROP | grep 100.64.0.0/10; then
              RULEFWD=$(iptables -L ts-forward --line-numbers | grep DROP | grep 100.64.0.0/10 | awk '{print $1}')
              RULEINP=$(iptables -L ts-input --line-numbers | grep DROP | grep 100.64.0.0/10 | awk '{print $1}')
              iptables -R ts-forward $RULEFWD -s 100.74.0.0/16 -o tailscale0 -j DROP
              iptables -R ts-input $RULEINP -s 100.74.0.0/16 -o tailscale0 -j DROP
            fi
  network:
    - name: "Reduce scope of traffic dropped by Tailscale to just the Tailscale ipPool"
      commands:
        - |
          systemctl enable tailscale-iptables-fix.service
          systemctl enable tailscale-iptables-fix.timer
          systemctl start tailscale-iptables-fix.timer
```

> This will ensure Tailscale does not drop traffic for IP ranges that it doesn't own. This is due to a known bug in Tailscale. Even though we restricted the IP Pool, Tailscale still puts in `iptables` rule on every node that drops unknown traffic from any address in the entire 100.64.0.0/10 range. 

Reviewer: If we can link to the bug, that'd be awesome.

Reviewer 2: Kevin said that he only managed to confirm it via email and this is not officially disclosed by tailscale
